semstr/conversion/ptb.py

```diff
@@ -0,0 +1,282 @@
+                            )
+                        # if it is a TExp object
+                        else:
+                            peers.append(head)
+                            head.next_sibling = tail
+                            tail = head
+                    stack.pop()
+
+                    if tx is None:
+                        tx = TExpr(None, node_id=node_id)
+
+                    tx.child_list = peers
+                    for child in tx.children():
+                        child.parent_node = tx
+                        # if the child is a terminal
+                        if isinstance(child.head, Leaf):
+                            self.node_ids_with_children.add(child.node_id)
+                            self.pending_nodes.append((child.parent_node.node_id, str(child.head.pos), child.node_id))
+                            self.node_ids_with_children.add(child.parent_node.node_id)
+                        # if the child is a regular node
+                        else:
+                            self.pending_nodes.append((child.parent_node.node_id, str(child.head), child.node_id))
+                            self.node_ids_with_children.add(child.parent_node.node_id)
+                            # node is root
+                            if child.parent_node.node_id is None:
+                                self.node_by_id[child.node_id] = None
+                    if not stack:
+                        self.node_by_id[node_id] = None
+                        yield tx
+                    else:
+                        stack.append(tx)
+                node_id += 1
+
+
+#######
+# Utils
+#######
+
+def gensym():
+    return object()
+
+
+##################
+# Lexer
+##################
+
+
+LPAREN_TOKEN = gensym()
+RPAREN_TOKEN = gensym()
+STRING_TOKEN = gensym()
+
+class Token(object):
+    _token_ids = {LPAREN_TOKEN:"(", RPAREN_TOKEN:")", STRING_TOKEN:"STRING"}
+
+    def __init__(self, token_id, value=None, lineno=None):
+        self.token_id = token_id
+        self.value = value
+        self.lineno = lineno
+
+    def __str__(self):
+        return "Token:'{tok}'{ln}".format(
+            tok=(self.value if self.value is not None else self._token_ids[self.token_id]),
+            ln=(':{}'.format(self.lineno) if self.lineno is not None else '')
+            )
+
+
+_token_pat = re.compile(r'\(|\)|[^()\s]+')
+def lex(line_or_lines):
+    """
+    Create a generator which returns tokens parsed from the input.
+    The input can be either a single string or a sequence of strings.
+    """
+
+    if isinstance(line_or_lines, str):
+        line_or_lines = [line_or_lines]
+
+    counter = 0
+    for n, line in enumerate(line_or_lines):
+        line.strip()
+        for m in _token_pat.finditer(line):
+            if m.group() == '(':
+                # if there is an extra '(' at the beginning of the sentence
+                if counter == 1:
```

Comment:
Shouldn't it be an error if there is an extra `(` at  the beginning?

Reply:
Some sentences starts with an extra `(` that plays the role of 'ROOT', so there's shouldn't be an error.  It doesn't get in count here in order to "ignore" the additional ROOT node that will appear if it will be taken into acount.

Reply:
Fine, but please document this in a comment.